### PR TITLE
Fixes GCS based ingestion

### PR DESCRIPTION
**Pull request by tejaswini-imply**
GCP allows buckets in GCS to contain underscores in their names. When this location is mapped to `java.net.URI`, `URI#host` comes out to be null as `URI` doesn't allow it to contain underscores which is being translated to bucket name hence mapping `URI#authority` to bucket name.

#### Fixed `CloudObjectLocation` constructor from `URI`

- Previously bucket variable of the `CloudObjectLocation` is being set by `uri.getHost()` value (but `java.net.URI` refuses to handle underscore in host since it's not a valid character and gcs allows underscores in bucket naming)

- This modification allows `uri.getAuthority()` to be mapped to bucket name which is allowing underscores to be present (Since all 3 S3, GCS, Azure Cloud buckets uris won't contain port or user information it is assumed that authority will return nothing more than the bucket name)

<hr>

##### Key changed/added classes in this PR
 * `CloudObjectLocation `

<hr>

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by kfaraz** [comment]
Thanks for the fix, @tejaswini-imply . 
Please add some unit tests to validate that this works in the following cases:
- with and without underscore
- GCS, S3 and other implementations

File: core/src/main/java/org/apache/druid/data/input/impl/CloudObjectLocation.java [hunk 1]
```diff
@@ -71,7 +71,7 @@ public CloudObjectLocation(@JsonProperty("bucket") String bucket, @JsonProperty(
 
   public CloudObjectLocation(URI uri)
   {
-    this(uri.getHost(), uri.getPath());
+    this(uri.getAuthority(), uri.getPath());
```
Comment (abhishekagarwal87) [inline core/src/main/java/org/apache/druid/data/input/impl/CloudObjectLocation.java:74]: should we fall back to authority only if getHost() returns null?

Comment (kfaraz) [reply core/src/main/java/org/apache/druid/data/input/impl/CloudObjectLocation.java:74]: Yeah, I would be more comfortable with that too.

Comment (tejaswini-imply) [reply core/src/main/java/org/apache/druid/data/input/impl/CloudObjectLocation.java:74]: That would work as well.

**Review by kfaraz — APPROVED** [review]
LGTM.

**Pull request merged**